File: p2.js
// ## Problem2 

// Longest Substring Without Repeating Characters(https://leetcode.com/problems/longest-substring-without-repeating-characters/)

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

const lengthOfLongestSubstring = function(s) {
    if(s.length === 0 || !s) return 0;
    let map = new Map(), max = 0, slow = 0;
    for(let i = 0; i < s.length; i++) {
        if(map.has(s[i])) {
            slow = Math.max(slow, map.get(s[i]));
        }
        max = Math.max(max, i - slow + 1);
        map.set(s[i], i+1);
    }
    return max;
};

File: p2.java
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach

class Solution {
    public int lengthOfLongestSubstring(String s) {
        //Hashmap to store the character and it's index
        HashMap<Character, Integer> map = new HashMap<>();
        int start = 0;
        int length = 0;
        //Go through all the characters in string and update start only when you fin a repeat occurence of a character
        for(int end=0; end<s.length(); end++){
            char c = s.charAt(end);
            if(map.containsKey(c)){
                int i = map.get(c);
                if(i>=start){
                    start = i+1;
                    
                }
            }
            map.put(c,end);
            //Get the length using start and end
            length = Math.max(length, end-start+1);
        }

        return length;
    }
}